code/modules/mob/living/carbon/xenomorph/abilities.dm

```diff
@@ -1371,76 +1362,154 @@ GLOBAL_LIST_INIT(pattern_images_list, list(
+	/// Holograms are used to show the pattern before placing it
+	var/list/holograms = list()
+	/// timerid before we cleanup the holograms
+	var/cleanup_timer
+	/// how far a hologram can visually move without being remade
+	var/move_range = 4
+	/// how long a hologram lasts without movement
+	var/cleanup_time = 4 SECONDS
 
 /datum/action/ability/activable/xeno/place_pattern/alternate_action_activate()
-	INVOKE_ASYNC(src, PROC_REF(selectpattern))
+	INVOKE_ASYNC(src, PROC_REF(select_pattern))
 
 /datum/action/ability/activable/xeno/place_pattern/give_action(mob/living/L)
 	. = ..()
 	//if its not prep, remove the ability instantly
 	if(!(SSmonitor.gamestate == SHUTTERS_CLOSED && CHECK_BITFIELD(SSticker.mode?.round_type_flags, MODE_ALLOW_XENO_QUICKBUILD) && SSresinshaping.active))
-		remove_action(xeno_owner)
+		remove_action(owner)
 	RegisterSignals(SSdcs, list(COMSIG_GLOB_OPEN_SHUTTERS_EARLY, COMSIG_GLOB_OPEN_TIMED_SHUTTERS_LATE,COMSIG_GLOB_OPEN_TIMED_SHUTTERS_XENO_HIVEMIND,COMSIG_GLOB_TADPOLE_LANDED_OUT_LZ,COMSIG_GLOB_TADPOLE_RAPPEL_DEPLOYED_OUT_LZ,COMSIG_GLOB_DROPPOD_LANDED), PROC_REF(toggle_off))
 
+///Seperate proc that calls remove_action, to block any signal shenanigans.
+/datum/action/ability/activable/xeno/place_pattern/proc/toggle_off()
+	SIGNAL_HANDLER
+	remove_action(owner)
+
 /datum/action/ability/activable/xeno/place_pattern/remove_action(mob/living/L)
 	. = ..()
 	UnregisterSignal(SSdcs, list(COMSIG_GLOB_OPEN_SHUTTERS_EARLY, COMSIG_GLOB_OPEN_TIMED_SHUTTERS_LATE,COMSIG_GLOB_OPEN_TIMED_SHUTTERS_XENO_HIVEMIND,COMSIG_GLOB_TADPOLE_LANDED_OUT_LZ,COMSIG_GLOB_TADPOLE_RAPPEL_DEPLOYED_OUT_LZ,COMSIG_GLOB_DROPPOD_LANDED))
 
-///Seperate proc that calls remove_action, to block any signal shenanigans.
-/datum/action/ability/activable/xeno/place_pattern/proc/toggle_off()
+/datum/action/ability/activable/xeno/place_pattern/on_selection()
+	RegisterSignal(owner, COMSIG_ATOM_MOUSE_ENTERED, PROC_REF(show_hologram_call))
+
+/datum/action/ability/activable/xeno/place_pattern/on_deselection()
+	UnregisterSignal(owner, COMSIG_ATOM_MOUSE_ENTERED)
+	cleanup_holograms()
+
+// don't slow down the other signals
+/datum/action/ability/activable/xeno/place_pattern/proc/show_hologram_call(mob/user, atom/target)
+	INVOKE_ASYNC(src, PROC_REF(show_hologram), user, target)
+
+/datum/action/ability/activable/xeno/place_pattern/proc/show_hologram(mob/user, atom/target)
 	SIGNAL_HANDLER
-	remove_action(xeno_owner)
+	var/list/target_turfs = get_target_turfs(target)
+	if(!length(target_turfs))
+		// no turfs? means the mouse is offscreen so we can cleanup
+		cleanup_holograms()
+		return
+	var/index = 1
+	var/create_new = should_replace_holograms(target_turfs)
+	if(create_new)
+		cleanup_holograms()
+	for(var/turf/target_turf in target_turfs)
+		if(create_new)
+			create_hologram(target_turf)
+			check_turf_validity(target_turf, holograms[index])
+		else
+			move_hologram(target_turf, holograms[index])
+			check_turf_validity(target_turf, holograms[index])
+		index++
```

Comment:
check_turf_validity could be once after the if else

Reply:
true